### PR TITLE
[expo-updates] add SelectionPolicyFilterAware

### PR DESCRIPTION
**Pull request by esamelson**
# Why

Support EAS Update's manifestFilters field, which is used for channels/rollouts. The client needs to be able to choose an update to launch, from a list of cached updates, without consulting the server--but using the stored value for `manifestFilters`. This PR adds a new `SelectionPolicy` on both platforms that takes any provided filters into account when selecting updates to load/launch/delete.

# How

Added `SelectionPolicyFilterAware`, which takes filters into account in the following ways:
- when selecting an update to launch, eliminates those which don't match the filters, then chooses the newest
- when selecting updates to delete, chooses to delete all updates that are older than the currently launched update, **except** the next newest update which matches the filters -- or, if no older updates match the filters, the next newest overall update
- when deciding whether to download a new update, uses the filters accordingly. There are some edge cases here which are better explained by reading the test cases. We will want to pass in the *new* filters here rather than any stored ones.

In this case, "match a filter" means that the manifest's `updateMetadata` object has a key that matches the filter **AND** the value doesn't match the expected one. In other words, if manifestFilters is the object `"channel": "rollout-1"`, an update with `updateMetadata: {channel: "default"}` would be filtered out but an update with `updateMetadata: {}` would **NOT** be filtered out.

# Test Plan

Added unit tests on both platforms for `SelectionPolicyFilterAware`.

# Todos

see https://www.notion.so/expo/EAS-Update-support-in-expo-updates-2178091dea9f4370aaf05f9a8ac328a6 . 
For this particular feature, next steps are to handle/parse the manifestFilters in the manifest response, persist them at the appropriate times, and then switch from `SelectionPolicyNewest` to `SelectionPolicyFilterAware`.

cc @jkhales

### CODE REVIEW

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+  @Test
+  public void testSelectUpdateToLaunch() {
+    // should pick the newest update that matches the manifest filters
+    UpdateEntity expected = updateRollout1;
+    UpdateEntity actual = selectionPolicy.selectUpdateToLaunch(Arrays.asList(updateDefault1, expected, updateDefault2), manifestFilters);
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void testSelectUpdatesToDelete_OlderMatching() {
```
Comment (ide) [inline packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:66]: It may make sense to change this test to "SecondNewestMatching" and to add updateRollout3 to show that matching manifests still are deleted if they aren't the newest or second newest.

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  @Test
+  public void testShouldLoadNewUpdate_NoneExistsMatchingCurrentFilters() {
+    // should choose to load a new update that doesn't match the manifest filters
+    // if no existing updates match the manifest filters
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateDefault2, null, manifestFilters);
+    Assert.assertTrue(actual);
+  }
+
+  @Test
+  public void testShouldLoadNewUpdate_NoneExistsMatchingNewFilters() {
```
Comment (ide) [inline packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:130]: Do we want this behavior? What are its consequences? If the server sends down a manifest that doesn't match the filters, do we want the app to update to it?

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+    // should choose to load a new update that doesn't match the manifest filters
+    // if no existing updates match the manifest filters
+    boolean actual = selectionPolicy.shouldLoadNewUpdate(updateDefault2, updateDefault1, manifestFilters);
+    Assert.assertTrue(actual);
+  }
+
+  @Test
+  public void testIsUpdateManifestFiltered_Nested() throws JSONException {
+    JSONObject nestedManifestFilters = new JSONObject("{\"nested.object.key\": \"different-value\"}");
```
Comment (ide) [inline packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:139]: Using dots here precludes us from having dots inside of updateMetadata field names -- we need to update the spec if we are going to introduce this restriction.

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+
+    // an update that has the matching key with a different value should be filtered out
+    Assert.assertTrue(selectionPolicy.isUpdateManifestFiltered(updateNested, nestedManifestFilters));
+
+    // an update that doesn't have the key should not be filtered out
+    Assert.assertFalse(selectionPolicy.isUpdateManifestFiltered(updateDefault1, nestedManifestFilters));
+  }
+}
```
Comment (ide) [inline packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:147]: May also want to add a test that shows that if a manifest doesn't have a field the filter looks for, it's implicitly counted as a match. (That said, do we want that behavior? If a developer says to match a given field and then they send down a manifest that omits the field altogether, I don't think they'd expect it to run.)

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+  UpdateEntity updateDefault1;
+  UpdateEntity updateDefault2;
+  UpdateEntity updateRollout1;
+  UpdateEntity updateRollout2;
+  UpdateEntity updateNested;
+
+  @Before
+  public void setup() throws JSONException {
+    manifestFilters = new JSONObject("{\"releaseName\": \"rollout\"}");
```
Comment (jkhales) [inline packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:34]:
```suggestion
    manifestFilters = new JSONObject("{\"branchName\": \"rollout\"}");
```

We are renaming `release` to `branch`

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java (hunk 2 above)
Comment (esamelson) [reply packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:130]: That's a good point -- discussed with @jkhales and we should return false here, essentially sanity checking that a manifest passes its own filters before loading it (and persisting the filters).

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java (hunk 3 above)
Comment (esamelson) [reply packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:139]: Good point. Discussed with @jkhales and we decided that we're just going to stick with non-nested key names for now, as this seems to cover all reasonable cases and is fully supported by the existing structured fields spec.

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java (hunk 4 above)
Comment (esamelson) [reply packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java:147]: Agree about adding a test. I'm pretty sure we do want this behavior; otherwise we'd need to start putting a `branchName` (or whatever other filters a user creates) into all embedded manifests in order for them not to be immediately filtered out.

(To be clear, I think we do want to allow users to specify `branchName` and other fields in embedded manifests, but we also don't want to force them to, which is what this "empty matches all" behavior allows.)

File: packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java (hunk 5 above)
Comment (esamelson) [reply packages/expo-updates/android/src/androidTest/java/expo/modules/updates/launcher/SelectionPolicyFilterAwareTest.java]: thanks, changed (although the nice thing about this implementation is that it's totally agnostic of the actual keys inside of manifestFilters!)

**Comment by esamelson** [comment]
Thanks for the feedback -- think I addressed all points.

Additionally, discussed with @ide and decided we want to **ignore** structured field value types that are not supported in JSON (e.g. binary) since there's no way we could have a manifest matching those values anyway; will add that functionality in a different PR since I think it makes most sense to do that when persisting the filters rather than when reading them.

**Review by ide — APPROVED** [review]
